File: src/backend/students/StudentServiceFunction/src/main/java/com/igp/studentservice/Handler.scala
package com.igp.studentservice

import com.amazonaws.services.lambda.runtime.{Context, RequestHandler}
import com.amazonaws.services.lambda.runtime.events.{APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent}
import com.igp.studentservice.service.{StudentService, StudentServiceImpl}
import com.igp.studentservice.util.{EnvironmentLogger, ResponseEntity}

/**
 * Trying scala for the handy pattern matching. Sorry if it's confusing you should never need to touch this class
 * anyways.
 */
class Handler extends RequestHandler[APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent] {

  override def handleRequest(event: APIGatewayProxyRequestEvent, context: Context): APIGatewayProxyResponseEvent = {
    EnvironmentLogger.log(event, context)

    routeRequest(event, context, new StudentServiceImpl).toApiGatewayProxyResponseEvent
  }

  def routeRequest(event: APIGatewayProxyRequestEvent, context: Context, service: StudentService): ResponseEntity[_] = event.getResource match {
    case "/students/basic/{id}" if event.getHttpMethod.equals("GET") => service.getBasicStudent(event, context)
    case "/students/create" if event.getHttpMethod.equals("POST") => service.createStudent(event, context)
    case "/students/delete/{id}" if event.getHttpMethod.equals("DELETE") => service.deleteStudent(event, context)
    case "/students/detailed/{id}" if event.getHttpMethod.equals("GET") => service.getDetailedStudent(event, context)
    case "/students/uni/{id}" if event.getHttpMethod.equals("GET") => service.getUniStudent(event, context)
    case "/students/update/{id}" if event.getHttpMethod.equals("POST") => service.updateStudent(event, context)
    case _ => ResponseEntity.notFound()
  }
}

File: src/backend/students/StudentServiceFunction/src/main/java/com/igp/studentservice/Handler.java
package com.igp.studentservice;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
import com.igp.studentservice.model.Student;
import com.igp.studentservice.util.AmazonExceptionHandler;
import com.igp.studentservice.util.pattern.PatternMatcher;
import com.igp.studentservice.service.StudentService;
import com.igp.studentservice.service.StudentServiceImpl;
import com.igp.studentservice.util.EnvironmentLogger;
import com.igp.studentservice.util.ResponseEntity;

public class Handler implements RequestHandler<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {

    @Override
    public APIGatewayProxyResponseEvent handleRequest(APIGatewayProxyRequestEvent event, Context context) {
        EnvironmentLogger.log(event, context);
        return routeRequest(event, context, new StudentServiceImpl()).toApiGatewayProxyResponseEvent();
    }

    private ResponseEntity<?> routeRequest(APIGatewayProxyRequestEvent event, Context context, StudentService service) {
        return AmazonExceptionHandler.handle(() -> {
            String resource = event.getResource();
            String httpMethod = event.getHttpMethod();

            if (resource.equalsIgnoreCase("/v1/students") && httpMethod.equals("POST")) {
                return ResponseEntity.ok(service.createStudent(event, context));
            } else if (resource.equalsIgnoreCase("/v1/students/{id}") && httpMethod.equals("GET")) {
                return ResponseEntity.ok(service.readStudent(event, context));
            } else if (resource.equalsIgnoreCase("/v1/students") && httpMethod.equals("UPDATE")) {
                return ResponseEntity.ok(service.updateStudent(event, context));
            } else if (resource.equalsIgnoreCase("/v1/students/{id}") && httpMethod.equals("DELETE")) {
                return ResponseEntity.ok(service.deleteStudent(event, context));
            } else {
                return ResponseEntity.notFound();
            }
        });
    }
}
